### PR TITLE
[LETS-41] Send transaction server confirmation that log was saved on page server

### PR DESCRIPTION
**Pull request by ribram**
http://jira.cubrid.org/browse/LETS-41

Page server sends a confirmation request to the active transaction server after saving log to disk. The first page server to active tran server request is added to this end.

**Request client/server**

- Do not move a request_server or request_client_server having started the server thread. The running thread keeps reference to old object.
- Change request_sync_send_queue to be constructed with a reference to request_client instead of moving the object.
- Adapt unit test with these changes.
- Fix broken header guard

**Page server**

- page_server changes:

  - Add page_server to active_tran_server request queue and autosend to handle sending requests from multiple sources.
  - Add push_active_tran_server_request() for sending requests to the active transaction server.

- logpb_flush_all_append_pages sends new nxio_lsa after saving log to disk.
- 

**Active transaction server**

- active_tran_server

  - Add m_ps_conn and avoid moving request client/server with a started thread.
  - Add handler for receiving saved LSA.

**Other**

- New system parameter to log commit confirmation procedures (sending saved LSA included)

**Testing**

Log messages regarding commit confirmation:
```
Page server:
Time: 03/19/21 00:55:57.671 - DEBUG *** file /home/remzi/sources/cubrid/src/transaction/log_page_buffer.c, line 3788
[COMMIT CONFIRM] Send saved LSA=109|9544.

Time: 03/19/21 00:55:59.666 - DEBUG *** file /home/remzi/sources/cubrid/src/transaction/log_prior_recv.cpp, line 95
[LOG_PRIOR_TRANSFER] Received list starting with lsa 109|9544. Message size = 138.

Time: 03/19/21 00:55:59.666 - DEBUG *** file /home/remzi/sources/cubrid/src/transaction/log_prior_send.cpp, line 39
[LOG PRIOR TRANSFER] Sending list starting with lsa 109|9544. Message size = 138.

Time: 03/19/21 00:55:59.672 - DEBUG *** file /home/remzi/sources/cubrid/src/transaction/log_page_buffer.c, line 3788
[COMMIT CONFIRM] Send saved LSA=109|9640.
```
```
Transaction server:
Time: 03/19/21 00:55:57.626 - DEBUG *** file /home/remzi/sources/cubrid/src/transaction/log_prior_send.cpp, line 39
[LOG PRIOR TRANSFER] Sending list starting with lsa 109|9448. Message size = 138.

Time: 03/19/21 00:55:57.671 - DEBUG *** file /home/remzi/sources/cubrid/src/server/active_tran_server.cpp, line 180
[COMMIT CONFIRM] Received LSA = 109|9544.

Time: 03/19/21 00:55:59.626 - DEBUG *** file /home/remzi/sources/cubrid/src/transaction/log_prior_send.cpp, line 39
[LOG PRIOR TRANSFER] Sending list starting with lsa 109|9544. Message size = 138.

Time: 03/19/21 00:55:59.673 - DEBUG *** file /home/remzi/sources/cubrid/src/server/active_tran_server.cpp, line 180
[COMMIT CONFIRM] Received LSA = 109|9640.
```

### CODE REVIEW

File: src/transaction/log_page_buffer.c [hunk 1]
```diff
@@ -3797,6 +3800,19 @@ logpb_flush_all_append_pages (THREAD_ENTRY * thread_p)
...
+#if defined (SERVER_MODE)
+  if (get_server_type () == SERVER_TYPE_PAGE)
+    {
+      log_lsa saved_lsa = log_Gl.append.get_nxio_lsa ();
+      if (prm_get_bool_value (PRM_ID_ER_LOG_COMMIT_CONFIRM))
+	{
+	  _er_log_debug (ARG_FILE_LINE, "[COMMIT CONFIRM] Send saved LSA=%lld|%d.\n", LSA_AS_ARGS (&saved_lsa));
+	}
+      std::string message (reinterpret_cast < const char *>(&saved_lsa), sizeof (saved_lsa));
+      ps_Gl.push_active_tran_server_request (ps_to_ats_request::SEND_SAVED_LSA, std::move (message));
```
Comment (cristiarg) [inline src/transaction/log_page_buffer.c:3812]: I find these names a bit misleading and hard to 'process'. In this case, is the `active_tran_server` the originator or the destination? I reckon it is the destination. Maybe introduce a "direction" particle like to/from in these names? `push_request_to_active_tran_server`?

File: src/communication/request_client_server.hpp [hunk 2]
```diff
@@ -246,6 +246,14 @@ namespace cubcomm
   {
   }
 
+  template <typename MsgId>
+  request_server<MsgId>::request_server (request_server &&other)
+  {
+    assert (!other.m_thread.joinable ());   // cannot move if thread is started
+    m_channel = std::move (other.m_channel);
+    m_request_handlers = std::move (other.m_request_handlers);
```
[the remark this answers is not in the document]
Comment (ribram) [reply src/communication/request_client_server.hpp:254]: it is implied.

File: src/server/active_tran_server.hpp [hunk 3]
```diff
@@ -43,9 +49,12 @@ class active_tran_server
   private:
     using page_server_request_autosend = cubcomm::request_queue_autosend<page_server_request_queue>;
 
+    void receive_saved_lsa (cubpacking::unpacker &upk);
+
     // communication with page server
     std::string m_ps_hostname;
-    int m_ps_port;
+    int m_ps_port = -1;
+    page_server_conn *m_ps_conn = nullptr;
```
[the remark this answers is not in the document]
Comment (ribram) [reply src/server/active_tran_server.hpp:57]: if they're replaced with unique_ptr, is there a guarantee that the objects are free in next order?

  1.  m_ps_request_autosend
  2. m_ps_request_queue
  3. m_ps_conn

File: src/transaction/log_page_buffer.c [hunk 4]
```diff
@@ -3797,6 +3800,19 @@ logpb_flush_all_append_pages (THREAD_ENTRY * thread_p)
     }
   flush_info->num_toflush = 0;
 
+#if defined (SERVER_MODE)
+  if (get_server_type () == SERVER_TYPE_PAGE)
+    {
+      log_lsa saved_lsa = log_Gl.append.get_nxio_lsa ();
```
[the remark this answers is not in the document]
Comment (ribram) [reply src/transaction/log_page_buffer.c]: done

File: src/transaction/log_page_buffer.c (hunk 1 above)
Comment (ribram) [reply src/transaction/log_page_buffer.c]: done

**Review by danilailie — APPROVED** [review]

**Review by cristiarg — APPROVED** [review]

File: src/server/active_tran_server.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (danilailie) [reply src/server/active_tran_server.hpp:57]: Maybe one object to contain them (and expose functionality for them) would make it easier to manage their lifetime and the order of their destruction.

**Comment by ribram** [comment]
@cristiarg removed the initializations

File: src/server/active_tran_server.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (cristiarg) [reply src/server/active_tran_server.hpp]: I agree. It can come with the next set of changes. Let's move on.

**Pull request merged**